### ISSUE TITLE
ENH: Support object arrays in matmul

### ISSUE THREAD

**Issue by homocomputeris** [numpy/numpy#11332]
It would be great to have the possibility to multiply object in a vertorized manner, e.g. a matrix by a vector of matrices or a matrix by a vector of SciPy's linear operator, without reshaping or comprehensions.

If it is "currently" not supported, is there a milestone for this feature?

**Comment by jaimefrio** [comment]
Thanks for the suggestion, but could you give some more concise example? It's unclear to me what exactly you are trying to do, how numpy fails to deliver that, and what you are hoping it would do...

**Comment by homocomputeris** [comment]
Sorry, I thought it was a known issue, and, if it's 'currently' unsupported, then there are plans to support it.

Well, for example, I need to calculate linear combinations of some linear operators (or any other objects that support multiplication and addition):
```python
import scipy.sparse.linalg as sla
import numpy as np


def mv(v):
    return 3 * v


dim = 2
factor = 5
shp = (dim, dim)
A = sla.LinearOperator(shp, matvec=mv)
As = np.array([A, A, A])
coefs = np.random.rand(dim * factor, len(As))

lincombs = coefs @ As

```
Now it gives
```
Traceback (most recent call last):
...
lincombs = coefs @ As
TypeError: Object arrays are not currently supported
```
So, one has to write something less concise and elegant:
```python
lincombs = np.sum(np.array(
    [[coef * lo for coef, lo in zip(row, As)] for row in coefs]), axis=1)
lincombs2 = np.sum(coefs * np.tile(As, (dim * factor, 1)), axis=1)
```

**Comment by eric-wieser** [comment]
Updated the title to reflect that you're specifically asking for `@`

**Comment by eric-wieser** [comment]
What would the semantics of `objA @ objB` be?

* objC[ind] = objA[ind], objB[ind]`  (elementwise matrix multiplication)
* objC[...,i,k] = sum(objA[...,i,j] * objB[...,j,k])` (generalization of the float `@`)
* objC[...,i,k] = sum(objA[...,i,j] @ objB[...,j,k])`  (works for block matrices)

**Comment by shoyer** [comment]
> `objC[...,i,k] = sum_j(objA[...,i,j] * objB[...,j,k])` (generalization of the float @)

I think this is all @homocomputeris needs here? (It's certainly the only one of these definitions that immediately makes sense to me.)

**Comment by homocomputeris** [comment]
I think the more direct approach would be to treat the object elements as scalars `objC[...,i,k] = sum_j(objA[...,i,j] * objB[...,j,k])`, and then apply general broadcasting `@` rules if necessary.
This way it fits some reasonable linear algebra assumptions. Obviously, if works if these object support multiplication, which also seems more reasonable that `@` support.

**Comment by eric-wieser** [comment]
Case 1 would be useful for a 1d array of polygons with differing numbers of vertices, where each polygon is a nx2 array of points - it would allow all the polygons to be transformed by the same matrix.

I think I agree that case 2 is the most reasonable, but it would be nice to have a way to do 1 and 3

**Comment by homocomputeris** [comment]
As far as I understand, `@` calls `matmul`. So, maybe, 1 and 3 can be implemented via different functions, that are not straightforward generalizations of matrix multiplication.

**Comment by seberg** [comment]
My gut feeling would be to go for option 2. and not think about it much. That this somewhat blocks the ability to do the other options, but I you should be able to replace floats for example with decimal and things should work as much as possible the same. And since scalars such as decimal probably do not implement `__matmul__` all others would break that.

**Comment by seberg** [comment]
Actually `np.dot` is already implemented like that anyway. The only reason it is probably not implement is that einsum does not have the corresponding loops is my guess.

**Comment by seberg** [comment]
I am not sure about how complicating it is. But my guess is that it is
not too bad. Implementing object arrays in einsum is probably straight
forward. I would not say simple, you will need to learn a bit about the
iterator for example and understand the einsum code partially.

OTOH, if you look through my "gist"s, there should be a pure python
implementation of it (and it is not long, could be a nice start to
understand the gist of C-einsum). The most difficult part might be to
just find the right place where/how to include it, with fast paths, the
C-einsum is a bit more complex (the object case may not need that much
though).

**Comment by mattip** [comment]
`matmul` should soon become a full-fledged `ufunc`, and stop using `einsum`. The change is waiting on PR #11175 and PR #11133.

**Comment by seberg** [comment]
Ah ok, right, then the einsum thing is separate from this really.

**Comment by EricCousineau-TRI** [comment]
Given that #11175 has landed, can I ask if there is anything else blocking this?

**Comment by EricCousineau-TRI** [comment]
Awesome! The only `dtype=object` stuff was a while back; can I ask for a good starting point, implementation- and testing-wise?

**Comment by eric-wieser** [comment]
Implementation wise, your best bet is to extract the integer matmul loop from the matmul template, and then replace c operators with pynumber calls

**Comment by hameerabbasi** [comment]
Closed by #13503?

**Comment by EricCousineau-TRI** [comment]
Ah, sweet! Thanks for commenting here!
(And sorry for dragging my feet!)
